qlib/TableMapper.qm

```diff
@@ -756,34 +773,62 @@ on_error table_mapper.rollback();
 
             @param rec a hash representing a single input record
 
-            @return if batch data was inserted then a hash (columns) of lists (row data) of all data inserted and potentially returned (in case of sequences) from the database server is returned; if constant mappings are used with batch data, then they are returned as single values assigned to the hash keys
+            @return if batch data was inserted (flushed) then a hash (columns) of lists (row data) of all data inserted and potentially returned (in case of sequences) from the database server is returned; if constant mappings are used with batch data, then they are returned as single values assigned to the hash keys; if nothing was flushed to the database in the call, Qore::NOTHING is returned
+
+            @note this function does not process hashes of lists (bulk format), it expects always single input record on the input
 
-            @throw MAPPER-BATCH-ERROR this exception is thrown if this method is called when the \c "unstable_input" option was given in the constructor
             @throw MISSING-INPUT a field marked mandatory is missing
             @throw STRING-TOO-LONG a field value exceeds the maximum value and the 'trunc' key is not set
             @throw INVALID-NUMBER the field is marked as numeric but the input value contains non-numeric data
+            @throw MAP-ERROR when the function is called wwith hash of lists (bulk) input
         */
         private *hash queueDataIntern(hash rec) {
             if (insert_block == 1) {
                 return insertRow(rec);
             }
 
+            *int rec_list_size = getRecListSize(rec);
+            if (exists rec_list_size) {
+                throw "MAP-ERROR", sprintf("TableMapper::queueDataIntern() does not "
+                        "accept hash of lists - this is likely some internal "
+                        "error. Got %y", rec);
+            }
+
             hash h = mapDataIntern(rec);
 
             # map record data to get the keys for the buffer
+            # FIXME ANOTHER problem: if we already get some thing in hbuf
```

Comment:
I'm not sure - the mapDataIntern returns a non-list for constant fields, right? By the following code in mapDataIntern:

```
            # copy all constant mappings to the output hash
            if (consth)
                h += consth;
```
so what happens to the constants then in hbuf - as I see it, for each call of queueDataIntern the constant is catenated? (this is what actually happens on postgres with one of the existing test). 

And if the constants are put into list, what happens when 1st queueDataIntern returns hash of lists like ("myconst" : 1, "normal_column" : (1,2,3)) and 2nd queueDataIntern reutrns ("myconst" : 1, "normal_column" : (4,5,6)).

then hbuf after 2nd call will be something like ("myconst" : (1,1), "normal_column" : (4,5,6)), or not?

I need to investigate more to understand it (and write more tests). Either way thanks a lot for your inputs.